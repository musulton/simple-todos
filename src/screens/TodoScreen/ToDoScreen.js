import React, {useState} from 'react';
import {ScrollView, StyleSheet, View,} from 'react-native';
import Heading from "../../shared/components/Heading";
import Input from "../../shared/components/Input";
import ToDoList from "./components/ToDoList";
import SubmitButton from "../../shared/components/SubmitButton";
import TabBar from "./components/TabBar";

let todoIndex = 0;

const ToDoScreen = () => {
    const [appState, setAppState] = useState({
        inputValue: '',
        todos: [],
        type: 'All'
    })

    const inputChange = (inputValue) => {
        setAppState({...appState, inputValue: inputValue});
    }

    const submitTodo = () => {
        if (appState.inputValue.match(/^\s*$/)) {
            return
        }
        const todo = {
            title: appState.inputValue,
            todoIndex,
            complete: false
        }
        todoIndex++
        const todos = [...appState.todos, todo]
        setAppState({...appState, inputValue: '', todos: todos});
    }

    const toggleComplete = (todoIndex) => {
        let {todos} = appState;
        todos.forEach((todo) => {
            if (todo.todoIndex === todoIndex) {
                todo.complete = !todo.complete;
            }
        })
        setAppState({...appState, todos: todos});
    }
    const deleteTodo = (todoIndex) => {
        let {todos} = appState;
        todos = todos.filter((todo) => todo.todoIndex !== todoIndex)
        setAppState({...appState, todos: todos});
    }

    const setType = (type) => {
        setAppState({...appState, type: type});
    }
    return (
        <View style={styles.container}>
            <Heading/>
            <Input inputValue={appState.inputValue} inputChange={(text) => inputChange(text)}/>
            <SubmitButton submitTodo={submitTodo}/>
            <ScrollView
                contentInsetAdjustmentBehavior="automatic" keyboardShouldPersistTaps="never" style={styles.content}>
                <ToDoList todos={appState.todos} toggleComplete={toggleComplete}
                          deleteTodo={deleteTodo} type={appState.type}/>

            </ScrollView>
            <TabBar type={appState.type} setType={setType}/>
        </View>
    );
};

const styles = StyleSheet.create({
    container: {
        flex: 1,
    },
    content: {
        flex: 1,
        paddingTop: 10,
    }
});

export default ToDoScreen;
