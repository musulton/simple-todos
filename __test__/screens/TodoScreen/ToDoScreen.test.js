import {render, fireEvent, screen} from "@testing-library/react-native";

import ToDoScreen from "../../../src/screens/TodoScreen/ToDoScreen";

jest.mock("react-redux", () => ({
    ...jest.requireActual("react-redux"),
    useSelector: jest.fn(),
    useDispatch: jest.fn(() => jest.fn())
}))

describe("ToDoScreen", () => {
    const mockOnSubmit = jest.fn();
    const mockOnDismissError = jest.fn();
    const mockOnLoad = jest.fn();
    const mockOnDelete = jest.fn();
    const mockOnCompleted = jest.fn();

    const onRefetchFn = expect.any(Function)

    const mockTodoService = jest.fn(() => ({
        onSubmitTodo: mockOnSubmit,
        onDismissError: mockOnDismissError,
        onLoadTodo: mockOnLoad,
        onDeleteTodo: mockOnDelete,
        onCompletedTodo: mockOnCompleted
    }))

    test("should return correctly", () => {
        render(<ToDoScreen todo={mockTodoService} />)

        expect(screen.toJSON()).toMatchSnapshot()
    })

    test("should call on submit service with params correctly", () => {
        render(<ToDoScreen todo={mockTodoService} />)

        fireEvent
            .changeText(screen.getByPlaceholderText("What needs to be done?"), "value")
        fireEvent.press(screen.getByText("Submit"))

        expect(mockOnSubmit).toHaveBeenCalledWith("value", onRefetchFn)
    })
})
