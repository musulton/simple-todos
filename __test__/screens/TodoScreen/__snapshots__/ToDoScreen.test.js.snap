// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDoScreen should return correctly 1`] = `
<View
  onLayout={[Function]}
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "flex": 1,
        "paddingTop": 10,
      }
    }
  >
    <View
      style={
        {
          "backgroundColor": "#ffffff",
          "elevation": 3,
          "marginLeft": 20,
          "marginRight": 20,
          "shadowColor": "#000000",
          "shadowOffset": {
            "height": 3,
            "width": 3,
          },
          "shadowOpacity": 0.2,
          "shadowRadius": 3,
        }
      }
    >
      <TextInput
        onChangeText={[Function]}
        placeholder="What needs to be done?"
        placeholderTextColor="#CACACA"
        selectionColor="#666666"
        style={
          {
            "backgroundColor": "#ffffff",
            "height": 60,
            "paddingLeft": 10,
            "paddingRight": 10,
          }
        }
        value=""
      />
    </View>
    <View
      style={
        {
          "alignItems": "flex-end",
        }
      }
    >
      <View
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#ffffff",
            "borderColor": "rgba(0,0,0,.1)",
            "borderWidth": 1,
            "height": 50,
            "justifyContent": "center",
            "marginRight": 20,
            "marginTop": 15,
            "paddingLeft": 20,
            "paddingRight": 20,
            "width": 200,
          }
        }
      >
        <Text
          style={
            {
              "color": "#666666",
              "fontWeight": "600",
            }
          }
        >
          Submit
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "flex": 1,
          "paddingTop": 10,
        }
      }
    >
      <RCTScrollView
        ListEmptyComponent={
          <EmptyState
            text="Data is still empty"
          />
        }
        contentContainerStyle={
          {
            "flex": 1,
          }
        }
        getItem={[Function]}
        getItemCount={[Function]}
        keyExtractor={[Function]}
        onContentSizeChange={[Function]}
        onLayout={[Function]}
        onMomentumScrollBegin={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEventThrottle={50}
        stickyHeaderIndices={[]}
        viewabilityConfigCallbackPairs={[]}
      >
        <View>
          <View
            style={
              {
                "alignItems": "center",
                "flex": 1,
                "justifyContent": "center",
              }
            }
          >
            <Text
              style={
                {
                  "fontSize": 16,
                  "fontWeight": "bold",
                }
              }
            >
              Data is still empty
            </Text>
          </View>
        </View>
      </RCTScrollView>
    </View>
    <View
      style={
        {
          "borderTopColor": "#dddddd",
          "borderTopWidth": 1,
          "flexDirection": "row",
          "height": 70,
        }
      }
    >
      <View
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
            },
            null,
            {
              "borderLeftColor": "#dddddd",
              "borderLeftWidth": 1,
            },
            null,
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "#777777",
                "fontSize": 16,
              },
              null,
            ]
          }
        >
          All
        </Text>
      </View>
      <View
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
            },
            null,
            {
              "borderLeftColor": "#dddddd",
              "borderLeftWidth": 1,
            },
            null,
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "#777777",
                "fontSize": 16,
              },
              null,
            ]
          }
        >
          Active
        </Text>
      </View>
      <View
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
            },
            null,
            {
              "borderLeftColor": "#dddddd",
              "borderLeftWidth": 1,
            },
            null,
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "#777777",
                "fontSize": 16,
              },
              null,
            ]
          }
        >
          Complete
        </Text>
      </View>
    </View>
  </View>
</View>
`;
